src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/ChainPal.BuildChain.cs

```diff
@@ -69,21 +73,73 @@ internal sealed partial class ChainPal : IDisposable, IChainPal
 
                             FILETIME ft = FILETIME.FromDateTime(verificationTime);
                             CertChainFlags flags = MapRevocationFlags(revocationMode, revocationFlag);
-                            ChainEngine chainEngine = useMachineContext ? ChainEngine.HCCE_LOCAL_MACHINE : ChainEngine.HCCE_CURRENT_USER;
-
                             SafeX509ChainHandle chain;
                             if (!Interop.crypt32.CertGetCertificateChain(chainEngine, certificatePal.CertContext, &ft, extraStoreHandle, ref chainPara, flags, IntPtr.Zero, out chain))
                                 return null;
+                            Interop.crypt32.CertFreeCertificateChainEngine(chainEngine);
                             return new ChainPal(chain);
                         }
                     }
                 }
             }
         }
 
-        private static SafeCertStoreHandle ConvertExtraStoreToSafeHandle(X509Certificate2Collection extraStore)
+        private static IntPtr GetChainEngine(
+            X509ChainTrustMode trustMode,
+            X509Certificate2Collection customTrustStore,
+            X509Certificate2Collection extraStore,
+            bool useMachineContext
+            )
+        {
+            IntPtr chainEngine;
+            if (trustMode == X509ChainTrustMode.CustomRootTrust)
+            {
+                CERT_CHAIN_ENGINE_CONFIG customChainEngine = new CERT_CHAIN_ENGINE_CONFIG();
+                customChainEngine.cbSize = Marshal.SizeOf<CERT_CHAIN_ENGINE_CONFIG>();
+                using (SafeCertStoreHandle customTrustStoreHandle = ConvertStoreToSafeHandle(customTrustStore))
+                {
+                    if (!customTrustStoreHandle.IsInvalid)
+                    {
+                        customChainEngine.hExclusiveRoot = customTrustStoreHandle.DangerousGetHandle();
+                    }
+
+                    Interop.crypt32.CertCreateCertificateChainEngine(customChainEngine, out chainEngine);
+                    AddDefaultCertificatesToStore(extraStore, useMachineContext);
+                }
+            }
+            else
+            {
+                chainEngine = useMachineContext ? (IntPtr)ChainEngine.HCCE_LOCAL_MACHINE : (IntPtr)ChainEngine.HCCE_CURRENT_USER;
+            }
+
+            return chainEngine;
+        }
+
+        private static void AddDefaultCertificatesToStore(X509Certificate2Collection collectionStore, bool useMachineContext)
+        {
+            StoreLocation storeLocation = useMachineContext ? StoreLocation.LocalMachine: StoreLocation.CurrentUser;
+            foreach (StoreName storeName in (StoreName[]) Enum.GetValues(typeof(StoreName)))
```

Comment:
This is expensive.  The array should be cached, and reflection should not be used to create it.